packages/cursorless-vscode/src/keyboard/grammar/grammarHelpers.ts

```diff
@@ -6,56 +7,123 @@ export type Unused = typeof UNUSED;
 
 /**
  * @param args The values output by the parser rule
- * @param argNames The keys to use for the payload
+ * @param argExtractors Extractors to get values for payload
  * @returns An object with the given keys mapped to the values at the same
  * positions in the parser rule's output
  */
-function constructPayload(args: any[], argNames: (string | Unused)[]) {
-  const arg: Record<string, any> = {};
-  for (let i = 0; i < argNames.length; i++) {
-    const name = argNames[i];
-    if (name === UNUSED) {
-      continue;
+function constructPayload<K>(
+  args: any[],
+  argExtractors: KeyboardArgExtractor<K>,
+): Record<keyof K, any> {
+  const arg: Partial<Record<keyof K, any>> = {};
+  for (const [key, value] of Object.entries(argExtractors)) {
+    if (value instanceof ArgPosition) {
+      arg[key as keyof K] = args[value.position];
+    } else {
+      arg[key as keyof K] = value;
     }
-    arg[name] = args[i];
   }
-  return arg;
+  return arg as Record<keyof K, any>;
 }
 
+class ArgPosition {
+  constructor(public position: number) {}
+}
+
+export const argPositions: Record<string, ArgPosition> = {
+  $0: new ArgPosition(0),
+  $1: new ArgPosition(1),
+  $2: new ArgPosition(2),
+  $3: new ArgPosition(3),
+  $4: new ArgPosition(4),
+  $5: new ArgPosition(5),
```

Comment:
I hope we never reach 3, much less 5.

Reply:
We already have an example of 4 (see relative exclusive), though we just use a plain old function there. Are you objecting to long rules or using this postprocessor for them?

Reply:
The latter. The one with four is readable, but I don’t think it would be nearly as readable with $1, $2, $3, $4.

Reply:
I think the first few are good.